src/adapter/src/coord/sequencer.rs

```diff
@@ -2176,50 +2174,77 @@ impl<S: Append + 'static> Coordinator<S> {
             .collect();
-        let (permutation, thinning) = permutation_for_arrangement(&key, typ.arity());
+        let (permutation, thinning): (HashMap<usize, usize>, _) =
+            permutation_for_arrangement(&key, typ.arity());
         // Two transient allocations. We could reclaim these if we don't use them, potentially.
         // TODO: reclaim transient identifiers in fast path cases.
         let view_id = self.allocate_transient_id()?;
         let index_id = self.allocate_transient_id()?;
-        // The assembled dataflow contains a view and an index of that view.
-        let mut dataflow = DataflowDesc::new(format!("temp-view-{}", view_id));
-        dataflow.set_as_of(Antichain::from_elem(timestamp));
-        let mut builder = self.dataflow_builder(compute_instance);
-        builder.import_view_into_dataflow(&view_id, &source, &mut dataflow)?;
-        for BuildDesc { plan, .. } in &mut dataflow.objects_to_build {
-            prep_relation_expr(
-                self.catalog.state(),
-                plan,
-                ExprPrepStyle::OneShot {
-                    logical_time: Some(timestamp),
-                    session,
-                },
-            )?;
-        }
-        dataflow.export_index(
-            index_id,
-            IndexDesc {
-                on_id: view_id,
-                key: key.clone(),
-            },
-            typ,
-        );
 
-        // Optimize the dataflow across views, and any other ways that appeal.
-        mz_transform::optimize_dataflow(&mut dataflow, &builder.index_oracle())?;
+        let dataflow = self.create_peek_dataflow(
+            view_id,
+            timestamp,
+            compute_instance,
+            session,
+            &source,
+            index_id,
+            &key,
+            typ.clone(),
+        )?;
 
-        // At this point, `dataflow_plan` contains our best optimized dataflow.
         // We will check the plan to see if there is a fast path to escape full dataflow construction.
         let peek_plan = self.create_peek_plan(
             dataflow,
             view_id,
             compute_instance,
             index_id,
-            key,
-            permutation,
+            key.clone(),
+            permutation.clone(),
             thinning.len(),
         )?;
 
+        // For slow path queries, search other clusters for an index. Don't
+        // look if the user has set a cluster replica though, because
+        // changing just the cluster session var could have a non-existent
+        // replica set.
+        if matches!(peek_plan, peek::PeekPlan::SlowPath(_))
+            && session.vars().cluster_replica().is_none()
+        {
+            let other_instances = self
+                .catalog
+                .compute_instances()
+                .filter(|i| i.id != compute_instance);
+            for instance in other_instances {
+                let dataflow = self.create_peek_dataflow(
```

Comment:
I'm a tiny bit concerned about the perf overhead of doing this for all slow path queries over all clusters. Any thoughts?